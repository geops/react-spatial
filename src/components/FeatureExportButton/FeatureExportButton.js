import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import KMLFormat from 'ol/format/KML';
import Layer from '../../layers/Layer';
import KML from '../../utils/KML';

const propTypes = {
  /**
   *  Children content of the Feature export button.
   */
  children: PropTypes.node,

  /**
   * Format to export features (function).
   * Supported formats: https://openlayers.org/en/latest/apidoc/module-ol_format_Feature-FeatureFormat.html
   */
  format: PropTypes.func,

  /**
   * An existing react-spatial Layer, using a valid ol.source.Vector
   */
  layer: PropTypes.instanceOf(Layer).isRequired,

  /**
   * Map projection.
   */
  projection: PropTypes.string,
};

const defaultProps = {
  children: null,
  format: KMLFormat,
  projection: 'EPSG:3857',
};

/**
 * This component displays a button to export geometry feature.<br>
 * Default export format is KML, which supported features' style export.<br>
 * Other formats do not always support style export (See specific format specs).
 */
class FeatureExportButton extends PureComponent {
  static createFeatureString(layer, projection, format) {
    if (format === KMLFormat) {
      return KML.writeFeatures(layer, projection);
    }

    // eslint-disable-next-line new-cap
    return new format().writeFeatures(layer.olLayer.getSource().getFeatures(), {
      featureProjection: projection,
    });
  }

  static exportFeatures(layer, projection, format) {
    const now = new Date()
      .toJSON()
      .slice(0, 20)
      .replace(/[.:T-]+/g, '');
    const featString = this.createFeatureString(layer, projection, format);

    const formatString = featString
      ? featString.match(/<(\w+)\s+\w+.*?>/)[1]
      : 'xml';

    const fileName = `exported_features_${now}.${formatString}`;
    const charset = document.characterSet || 'UTF-8';
    const type = `${
      formatString === 'kml'
        ? 'data:application/vnd.google-earth.kml+xml'
        : 'data:text/xml'
    };charset=${charset}`;

    if (featString) {
      if (window.navigator.msSaveBlob) {
        // ie 11 and higher
        window.navigator.msSaveBlob(new Blob([featString], { type }), fileName);
      } else {
        const link = document.createElement('a');
        link.download = fileName;
        link.href = `${type},${encodeURIComponent(featString)}`;
        link.click();
      }
    }
  }

  render() {
    const { children, layer, projection, format, ...other } = this.props;

    return (
      <div
        role="button"
        className="rs-feature-export-button"
        tabIndex={0}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...other}
        onClick={() =>
          FeatureExportButton.exportFeatures(layer, projection, format)
        }
        onKeyPress={(evt) =>
          evt.which === 13 &&
          FeatureExportButton.exportFeatures(layer, projection, format)
        }
      >
        {children}
      </div>
    );
  }
}

FeatureExportButton.propTypes = propTypes;
FeatureExportButton.defaultProps = defaultProps;

export default FeatureExportButton;
