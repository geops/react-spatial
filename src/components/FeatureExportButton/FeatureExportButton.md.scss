.tm-feature-export-example {
  display: flex;
  padding: 15px 0;

  .tm-feature-export {
    margin-right: 15px;
    width: 150px;
    padding: 5px;
    border-radius: 10px;
    border: 2px solid $gray-lighter;

    &:hover {
      background-color: $gray-lighter;
    }
  }
}
