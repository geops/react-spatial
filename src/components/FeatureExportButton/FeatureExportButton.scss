.tm-feature-export {
  cursor: pointer;
  height: $btn-size-base;
  width: $btn-size-base;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 10px;
  border: 2px solid $gray-lighter;
}
