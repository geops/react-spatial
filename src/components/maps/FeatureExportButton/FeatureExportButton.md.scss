.rs-feature-export-example {
  .rs-feature-export-button {
    display: flex;
    align-items: center;
    justify-content: center;
    margin: 15px;
    width: 150px;
    height: 50px;
    border-radius: 10px;
    border: 2px solid $gray-lighter;

    &:hover {
      background-color: $gray-lighter;
    }
  }
}
