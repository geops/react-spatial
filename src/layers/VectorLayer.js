import { unByKey } from 'ol/Observable';
import Layer from './Layer';

/**
 * A class representing vector layer to display on BasicMap
 * @class
 * @inheritDoc
 * @param {Object} [options]
 */
class VectorLayer extends Layer {
  constructor(options = {}) {
    super(options);

    this.hitTolerance = options.hitTolerance || 5;

    // Array of click callbacks
    this.clickCallbacks = [];

    // Add click callback
    if (options.onClick) {
      this.onClick(options.onClick);
    }
  }

  /**
   * Listens to click events on the layer.
   * @param {function} callback Callback function, called with the clicked
   *   features (https://openlayers.org/en/latest/apidoc/module-ol_Feature.html),
   *   the layer instance and the click coordinate.
   */
  onClick(callback) {
    if (typeof callback === 'function') {
      this.clickCallbacks.push(callback);
    } else {
      throw new Error('callback must be of type function.');
    }
  }

  /**
   * Request feature information for a given coordinate.
   * @param {ol.Coordinate} coordinate {@link https://openlayers.org/en/latest/apidoc/module-ol_coordinate.html ol/Coordinate} to request the information at.
   * @returns {Promise<Object>} Promise with features, layer and coordinate
   *  or null if no feature was hit.
   * eslint-disable-next-line class-methods-use-this
   */
  getFeatureInfoAtCoordinate(coordinate) {
    let features = [];

    if (this.map) {
      const pixel = this.map.getPixelFromCoordinate(coordinate);
      features = this.map.getFeaturesAtPixel(pixel, {
        layerFilter: l => l === this.olLayer,
        hitTolerance: this.hitTolerance,
      });
    }

    return Promise.resolve({
      features,
      layer: this,
      coordinate,
    });
  }

  /**
   * Initialize the layer and listen to feature clicks.
   * @param {ol.map} map {@link https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html ol/Map}
   */
  init(map) {
    super.init(map);

    if (!this.map) {
      return;
    }

    // Listen to click events
    this.singleClickRef = this.map.on('singleclick', e => {
      if (!this.clickCallbacks.length) {
        return;
      }

      this.getFeatureInfoAtCoordinate(e.coordinate)
        .then(d => this.callClickCallbacks(d.features, d.layer, d.coordinate))
        .catch(() => this.callClickCallbacks([], this, e.coordinate));
    });
  }

  /**
   * Call click callbacks with given parameters.
   * This is done in a separate function for being able to modify the response.
   * @private
   */
  callClickCallbacks(features, layer, coordinate) {
    this.clickCallbacks.forEach(c => c(features, layer, coordinate));
  }

  /**
   * Terminate what was initialized in init function. Remove layer, events...
   */
  terminate() {
    super.terminate();
    if (this.singleClickRef) {
      unByKey(this.singleClickRef);
    }
  }
}

export default VectorLayer;
