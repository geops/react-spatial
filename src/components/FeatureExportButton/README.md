#

This demonstrates the use of FeatureExportButton.

```jsx
import React from 'react';
import BasicMap from 'react-spatial/components/BasicMap';
import Layer from 'react-spatial/Layer';
import VectorLayer from 'react-spatial/VectorLayer';
import OLMap from 'ol/Map';
import VectorSource from 'ol/source/Vector';
import TileLayer from 'ol/layer/Tile';
import Feature from 'ol/Feature';
import Point from 'ol/geom/Point';
import Style from 'ol/style/Style';
import Icon from 'ol/style/Icon';
import GPX from 'ol/format/GPX';
import OSM, {ATTRIBUTION} from 'ol/source/OSM.js';

import FeatureExportButton from 'react-spatial/components/FeatureExportButton';

class FeatureExportButtonExample extends React.Component {
  constructor(props) {
    super(props);
    this.map = new OLMap();

    const featStyle = new Style({
      image: new Icon({
        anchor: [0.5, 46],
        anchorXUnits: 'fraction',
        anchorYUnits: 'pixels',
        src: 'https://openlayers.org/en/latest/examples/data/icon.png',
      }),
    });

    this.layers = [
      new Layer({
        olLayer: new TileLayer({
          source: new OSM(),
        }),
      }),
      new VectorLayer({
        name: 'ExportLayer',
        source: new VectorSource({
          features: [
            new Feature({
              geometry: new Point([819103.972418, 6120013.078324]),
            }),
            new Feature({
              geometry: new Point([873838.856313, 6106009.575876]),
            }),
          ],
        }),
      }),
    ];

    // Need to assign a style to each Feature.
    this.layers[1].olLayer.getSource().forEachFeature(f => {
      f.setStyle(featStyle);
    });
  }

  render() {
    return (
      <div>
        <BasicMap
          center={[843119.531243, 6111943.000197]}
          zoom={9}
          map={this.map}
          layers={this.layers}
        />
        <div className="tm-feature-export-example">
          <FeatureExportButton layer={this.layers[1]}>
            Export as KML
          </FeatureExportButton>
          <FeatureExportButton
            format={GPX}
            layer={this.layers[1]}
          >
            Export as GPX
          </FeatureExportButton>
        </div>
      </div>
    );
  }
}

<FeatureExportButtonExample />;
```
