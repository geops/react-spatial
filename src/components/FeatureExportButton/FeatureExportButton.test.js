import 'jest-canvas-mock';
import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { TiImage } from 'react-icons/ti';
import GPX from 'ol/format/GPX';
import VectorSource from 'ol/source/Vector';
import Style from 'ol/style/Style';
import Icon from 'ol/style/Icon';
import Text from 'ol/style/Text';
import Fill from 'ol/style/Fill';
import Stroke from 'ol/style/Stroke';
import Circle from 'ol/geom/Circle';
import Point from 'ol/geom/Point';
import Feature from 'ol/Feature';
import VectorLayer from '../../VectorLayer';

import FeatureExportButton from '.';

configure({ adapter: new Adapter() });

describe('FeatureExportButton', () => {
  const layer = new VectorLayer({
    source: new VectorSource({
      features: [
        new Feature({
          geometry: new Point([819103.972418, 6120013.078324]),
        }),
      ],
    }),
  });

  test('should match snapshot.', () => {
    const component = renderer.create(
      <FeatureExportButton
        title="FeatureExport"
        layer={layer}
        className="tm-feature-export-example"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should match snapshot with children passed.', () => {
    const component = renderer.create(
      <FeatureExportButton layer={layer}>
        <TiImage />
      </FeatureExportButton>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('triggers onClick', () => {
    // Use library 'jest-environment-jsdom-fourteen'
    // to allow jest test of format.writeFeatures(feats).
    const renderLayer = featNumbers => {
      const featsArray = [];
      for (let i = 0; i < featNumbers; i += 1) {
        featsArray.push(
          new Feature({
            geometry: new Point([819103.972418, 6120013.078324]),
          }),
        );
      }

      return new VectorLayer({
        name: 'ExportLayer',
        source: new VectorSource({
          features: featsArray,
        }),
      });
    };

    const iconLayer = renderLayer(1);

    test('should be trigger click function.', () => {
      const wrapper = shallow(<FeatureExportButton layer={iconLayer} />);
      const spy = jest.spyOn(FeatureExportButton.prototype, 'exportFeature');

      wrapper.find('.tm-feature-export').simulate('click');
      expect(spy).toHaveBeenCalledTimes(1);
    });

    test('should export kml by default.', () => {
      const wrapper = mount(<FeatureExportButton layer={iconLayer} />);
      const exportString = wrapper.instance().createFeatureString(iconLayer);
      expect(exportString.match(/<(\w+)\s+\w+.*?>/)[1]).toBe('kml');
    });

    test('should export gpx format.', () => {
      const wrapper = mount(
        <FeatureExportButton format={GPX} layer={iconLayer} />,
      );
      const exportString = wrapper.instance().createFeatureString(iconLayer);
      expect(exportString.match(/<(\w+)\s+\w+.*?>/)[1]).toBe('gpx');
    });

    const iconStyle = new Style({
      image: new Icon({
        anchor: [0.5, 46],
        anchorXUnits: 'fraction',
        anchorYUnits: 'pixels',
        src: 'https://openlayers.org/en/latest/examples/data/icon.png',
      }),
    });

    iconLayer.olLayer.getSource().forEachFeature(f => {
      f.setStyle(iconStyle);
    });

    test('should export icon style in kml.', () => {
      const wrapper = mount(<FeatureExportButton layer={iconLayer} />);
      const exportString = wrapper.instance().createFeatureString(iconLayer);
      const expectedString =
        '<Style><IconStyle><Icon><href>' +
        'https://openlayers.org/en/latest/examples/data/icon.png' +
        '</href></Icon></IconStyle></Style>';
      expect(exportString.match(/<Style>(.*?)<\/Style>/g)[0]).toBe(
        expectedString,
      );
    });

    test('should export with layer name.', () => {
      const namedlayer = renderLayer(1);
      const wrapper = mount(<FeatureExportButton layer={namedlayer} />);
      const exportString = wrapper.instance().createFeatureString(namedlayer);
      const expectedname = '<name>ExportLayer</name>';
      expect(exportString.match(/<name>(.*?)<\/name>/g)[0]).toBe(expectedname);
    });

    test('should export without layer name.', () => {
      const unnamedlayer = renderLayer(1);
      delete unnamedlayer.name;
      const wrapper = mount(<FeatureExportButton layer={unnamedlayer} />);
      const exportString = wrapper.instance().createFeatureString(unnamedlayer);
      expect(/<name>/g.test(exportString)).toBe(false);
    });

    test('should export text style in kml.', () => {
      const textlayer = renderLayer(2);

      const textStyle = new Style({
        text: new Text({
          font: 'normal 16px Helvetica',
          stroke: new Stroke({
            color: [255, 255, 255, 1],
            width: 3,
          }),
        }),
      });

      textlayer.olLayer.getSource().forEachFeature(f => {
        f.setStyle(textStyle);
      });

      const wrapper = mount(<FeatureExportButton layer={textlayer} />);
      const exportString = wrapper.instance().createFeatureString(textlayer);
      const expectedStyle =
        '<Style><IconStyle><scale>0</scale></IconStyle>' +
        '<LabelStyle><color>ff333333</color></LabelStyle></Style>';
      expect(exportString.match(/<Style>(.*?)<\/Style>/g)[0]).toBe(
        expectedStyle,
      );
    });

    test("should not export 'Cirle geom' (kml unsupported).", () => {
      const circleLayer = new VectorLayer({
        name: 'ExportLayer',
        source: new VectorSource({
          features: [
            new Feature({
              geometry: new Circle({
                center: [843119.531243, 6111943.000197],
                radius: 1000,
              }),
            }),
          ],
        }),
      });

      const circleStyle = new Style({
        stroke: new Stroke({
          color: 'red',
          width: 3,
        }),
        fill: new Fill({
          color: 'rgba(255, 0, 0, 0.1)',
        }),
      });

      circleLayer.olLayer.getSource().forEachFeature(f => {
        f.setStyle(circleStyle);
      });

      const wrapper = mount(<FeatureExportButton layer={circleLayer} />);
      const exportString = wrapper.instance().createFeatureString(circleLayer);
      expect(exportString).toBe(undefined);
    });
  });
});
