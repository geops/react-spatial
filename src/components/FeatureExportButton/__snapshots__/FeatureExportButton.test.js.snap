// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureExportButton should match snapshot with children passed. 1`] = `
<div
  className="tm-button tm-feature-export"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <svg
    baseProfile="tiny"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    version="1.2"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="8.5"
      cy="8.501"
      r="2.5"
    />
    <path
      d="M16 10c-2 0-3 3-4.5 3s-1.499-1-3.5-1c-2 0-3.001 4-3.001 4h14.001s-1-6-3-6zM20 3h-16c-1.103 0-2 .897-2 2v12c0 1.103.897 2 2 2h16c1.103 0 2-.897 2-2v-12c0-1.103-.897-2-2-2zm0 14h-16v-12h16v12z"
    />
  </svg>
</div>
`;

exports[`FeatureExportButton should match snapshot. 1`] = `
<div
  aria-label="FeatureExport"
  className="tm-feature-export-example"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
  title="FeatureExport"
>
  <svg
    fill="currentColor"
    focusable={false}
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 640 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M537.6 226.6c4.1-10.7 6.4-22.4 6.4-34.6 0-53-43-96-96-96-19.7 0-38.1 6-53.3 16.2C367 64.2 315.3 32 256 32c-88.4 0-160 71.6-160 160 0 2.7.1 5.4.2 8.1C40.2 219.8 0 273.2 0 336c0 79.5 64.5 144 144 144h368c70.7 0 128-57.3 128-128 0-61.9-44-113.6-102.4-125.4zm-132.9 88.7L299.3 420.7c-6.2 6.2-16.4 6.2-22.6 0L171.3 315.3c-10.1-10.1-2.9-27.3 11.3-27.3H248V176c0-8.8 7.2-16 16-16h48c8.8 0 16 7.2 16 16v112h65.4c14.2 0 21.4 17.2 11.3 27.3z"
    />
  </svg>
</div>
`;
